LanguageFeatures/Patterns/type_inference_A33_t02.dart

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// If p with required type T is in an irrefutable context:
+/// - It is a compile-time error if M is not assignable to T.
+/// - Else if M is not a subtype of T then an implicit coercion or cast is
+///   inserted before the pattern binds the value, tests the value's type,
+///   destructures the value, or invokes a function with the value as a target
+///   or argument.
+///
+/// @description Check that if `M` is not a subtype of `T` then generic function
+/// instantiation is performed
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/expect.dart";
+
+T foo<T>(T t) => t;
+
+main() {
+  var list = [foo];
+  var [int Function(int) f1] = list;
```

Comment:
You could say that this is silly, but I think `var <int Function(int)>[f1] = ...` is better. The point is that we're consistently using the same mechanism to obtain a firmly and explicitly typed pattern.

Note that the type argument of the list pattern could be `dynamic`, so that's one thing which is less firm with `[int Function(int) f1]` than it is with `<int Function(int)>[f1]`.

We already have tests confirming that the declared type of `f1` in this case is `int Function(int)`, so it should not be necessary to declare the type of `f1` explicitly when the type argument of the list pattern is given explicitly.

Reply:
Changed. Please note that updated test fails in bot analyzer and dartk
```
unexpected analysis errors in type_inference_A33_t02.dart:
- Line 27, column 7: COMPILE_TIME_ERROR.PATTERN_TYPE_MISMATCH_IN_IRREFUTABLE_CONTEXT
  The matched value of type 'List<T Function<T>(T)>' isn't assignable to the required type 'List<int Function(int)>'.
```

Reply:
Of course, that's an error! ;-)

So we need to consider one more time how we can test the matched value type against the required type of the pattern. I think it isn't possible to obtain the generic function instantiation for the list pattern as such: We don't have a list-level coercion (like, "transform the elements one at a time").

This means that we'll just have to give up on the list (which means that it doesn't actually matter that the thing we're doing happens in a list, we could just as well have done it at the top level or in a record, etc.).

```dart
var <T Function<T>(T)>[int Function(int) f1] = list;
```

This will actually yield a situation where the matched value type is `T Function<T>(T)` and the pattern required type is `int Function(int)`, and I expect the spec wording to cover this case (so we should get a coercion, not a compile-time error).

Again, we could leave out the type argument on the pattern, but it's more readable to have it because it shows exactly the thing which is being tested: The relationship between the list pattern type argument and the declared type of the element, and the actions taken when they differ in this specific way.

Fun thing to try: :smile_cat:

```dart
var <T Function<T>(T)>[int Function(int) f1int && String Function(String) f1String] = list;
```